gme/ext/emu2413.c

```diff
@@ -1,1618 +1,1231 @@
-      perc -= calc_slot_cym (CAR(opll,8), MOD(opll,7)->pgout);
-  }
- } // end if (! opll->vrc7_mode)
+  opll->rhythm_mode = 0;
+  opll->slot_key_status = 0;
+  opll->eg_counter = 0;
 
-  out = inst + (perc << 1);
-  return (e_int16) out;
-}
+  reset_rate_conversion_params(opll);
 
-#ifdef EMU2413_COMPACTION
-e_int16
-OPLL_calc (OPLL * opll)
-{
-  return calc (opll);
-}
-#else
-e_int16
-OPLL_calc (OPLL * opll)
-{
-  if (!opll->quality)
-    return calc (opll);
-
-  while (opll->realstep > opll->oplltime)
-  {
-    opll->oplltime += opll->opllstep;
-    opll->prev = opll->next;
-    opll->next = calc (opll);
+  for (i = 0; i < 18; i++)
+    reset_slot(&opll->slot[i], i);
+
+  for (i = 0; i < 9; i++) {
+    set_patch(opll, i, 0);
   }
 
-  opll->oplltime -= opll->realstep;
-  opll->out = (e_int16) (((double) opll->next * (opll->opllstep - opll->oplltime)
-                          + (double) opll->prev * opll->oplltime) / opll->opllstep);
+  for (i = 0; i < 0x40; i++)
+    OPLL_writeReg(opll, i, 0);
+
+  for (i = 0; i < 15; i++) {
+    opll->pan[i] = 3;
+    opll->pan_fine[i][1] = opll->pan_fine[i][0] = 1.0f;
+  }
 
-  return (e_int16) opll->out;
+  for (i = 0; i < 14; i++) {
+    opll->ch_out[i] = 0;
+  }
 }
-#endif
 
-/*e_uint32
-OPLL_setMask (OPLL * opll, e_uint32 mask)
-{
-  e_uint32 ret;
+void OPLL_forceRefresh(OPLL *opll) {
+  int i;
 
-  if (opll)
-  {
-    ret = opll->mask;
-    opll->mask = mask;
-    return ret;
-  }
-  else
-    return 0;
-}*/
-
-void OPLL_SetMuteMask(OPLL* opll, e_uint32 MuteMask)
-{
-	unsigned char CurChn;
-	e_uint32 ChnMsk;
-
-	for (CurChn = 0; CurChn < 14; CurChn ++)
-	{
-		if (CurChn < 9)
-		{
-			ChnMsk = OPLL_MASK_CH(CurChn);
-		}
-		else
-		{
-			switch(CurChn)
-			{
-			case 9:
-				ChnMsk = OPLL_MASK_BD;
-				break;
-			case 10:
-				ChnMsk = OPLL_MASK_SD;
-				break;
-			case 11:
-				ChnMsk = OPLL_MASK_TOM;
-				break;
-			case 12:
-				ChnMsk = OPLL_MASK_CYM;
-				break;
-			case 13:
-				ChnMsk = OPLL_MASK_HH;
-				break;
-			default:
-				ChnMsk = 0;
-				break;
-			}
-		}
-		if ((MuteMask >> CurChn) & 0x01)
-			opll->mask |= ChnMsk;
-		else
-			opll->mask &= ~ChnMsk;
-	}
-
-	return;
-}
-
-/*e_uint32
-OPLL_toggleMask (OPLL * opll, e_uint32 mask)
-{
-  e_uint32 ret;
-
-  if (opll)
-  {
-    ret = opll->mask;
-    opll->mask ^= mask;
-    return ret;
+  if (opll == NULL)
+    return;
+
+  for (i = 0; i < 9; i++) {
+    set_patch(opll, i, opll->patch_number[i]);
   }
-  else
-    return 0;
-}*/
 
-void OPLL_SetChipMode(OPLL* opll, e_uint8 Mode)
-{
-	// Enable/Disable VRC7 Mode (with only 6 instead of 9 channels and no rhythm part)
-	opll->vrc7_mode = Mode;
+  for (i = 0; i < 18; i++) {
+    request_update(&opll->slot[i], UPDATE_ALL);
+  }
+}
 
-	return;
+void OPLL_setRate(OPLL *opll, uint32_t rate) {
+  opll->rate = rate;
+  reset_rate_conversion_params(opll);
 }
 
-/****************************************************
+void OPLL_setQuality(OPLL *, uint8_t) {}
```

Comment:
Not all compilers like this:

```
./gme/ext/emu2413.c: In function 'OPLL_setQuality':
./gme/ext/emu2413.c:1201:1: error: parameter name omitted
```

Suggesting this:

```diff
--- gme/ext/emu2413.c~
+++ gme/ext/emu2413.c
@@ -1201 +1201 @@
-void OPLL_setQuality(OPLL *, uint8_t) {}
+void OPLL_setQuality(OPLL *opll, uint8_t q) { (void)opll;(void)q; }
```

Reply:
Oh right, that's a c++ thing